File: plugins/claude-organizer/skills/implement/scripts/attach-commit.py
#!/usr/bin/env python3
"""Capture a commit's diff and attach it to its card via the REST API.

    python3 scripts/attach-commit.py <sha> [CO-N]

The card key is parsed from the commit message (``feat(...): … (CO-12)``) unless
passed explicitly. The diff goes straight from ``git`` to the API over HTTP — it
never passes through an AI context (no MCP, no tokens spent reading it).

Standard library only; a Node twin lives at scripts/attach-commit.mjs for
machines without Python. Keep the two in sync.

Config: CO_API_URL (default http://127.0.0.1:4400).
"""

import json
import os
import re
import subprocess
import sys
import urllib.error
import urllib.request

API_URL = os.environ.get("CO_API_URL", "http://127.0.0.1:4400").rstrip("/")

# Card-scoped token minted by the MCP (issue_commit_token); only needed when the
# API has auth on. Absent in sem-auth mode — then no extra header is sent.
COMMIT_TOKEN = os.environ.get("CO_COMMIT_TOKEN")


def with_token(headers):
    if COMMIT_TOKEN:
        return {**headers, "X-CO-Commit-Token": COMMIT_TOKEN}
    return headers

# Files whose body is noise: store the header + a note instead of the patch.
LOCKFILES = {
    "pnpm-lock.yaml",
    "package-lock.json",
    "yarn.lock",
    "npm-shrinkwrap.json",
    "bun.lockb",
}
# Per-file line cap; beyond it the patch is truncated with a note.
MAX_LINES_PER_FILE = 1000

# `CO-12`, `ABC-7` — an uppercase prefix, a dash, digits.
KEY_RE = re.compile(r"\b([A-Z][A-Z0-9]*-\d+)\b")


def fail(msg):
    print(f"✗ {msg}", file=sys.stderr)
    sys.exit(1)


def git(args):
    try:
        return subprocess.run(
            ["git", *args],
            capture_output=True,
            text=True,
            check=True,
        ).stdout
    except subprocess.CalledProcessError as err:
        first = (err.stderr or "").split("\n", 1)[0]
        fail(f"git {' '.join(args)} failed: {first}")


def parse_key(message):
    """Prefer the last key on the subject line, else the first anywhere."""
    subject = message.split("\n", 1)[0]
    on_subject = KEY_RE.findall(subject)
    if on_subject:
        return on_subject[-1]
    anywhere = KEY_RE.findall(message)
    return anywhere[0] if anywhere else None


def is_ignored(path):
    base = path.rsplit("/", 1)[-1]
    if base in LOCKFILES:
        return True
    if re.search(r"\.min\.[^.]+$", base):
        return True
    if path.startswith("dist/") or "/dist/" in path:
        return True
    return False


def section_path(lines):
    for l in lines:
        if l.startswith("+++ b/"):
            return l[6:]
    for l in lines:
        if l.startswith("--- a/"):
            return l[6:]
    m = re.match(r"^diff --git a/(.+) b/(.+)$", lines[0])
    return m.group(2) if m else ""


def prune_diff(patch):
    """Split the unified patch per file and apply the safeguards."""
    if not patch.strip():
        return ""
    sections = []
    current = None
    for line in patch.split("\n"):
        if line.startswith("diff --git "):
            if current:
                sections.append(current)
            current = [line]
        elif current is not None:
            current.append(line)
    if current:
        sections.append(current)

    out = []
    for lines in sections:
        path = section_path(lines)
        is_binary = any(
            re.match(r"^Binary files .* differ$", l) or l.startswith("GIT binary patch")
            for l in lines
        )
        if is_binary:
            out.append("\n".join(lines))
        elif is_ignored(path):
            out.append(f"{lines[0]}\n# diff omitted (lockfile/generated)")
        elif len(lines) > MAX_LINES_PER_FILE:
            omitted = len(lines) - MAX_LINES_PER_FILE
            out.append(
                "\n".join(lines[:MAX_LINES_PER_FILE])
                + f"\n# ... (truncated: {omitted} lines omitted)"
            )
        else:
            out.append("\n".join(lines))
    return "\n".join(out)


def human_bytes(n):
    if n < 1024:
        return f"{n} B"
    if n < 1024 * 1024:
        return f"{n / 1024:.1f} KB"
    return f"{n / 1024 / 1024:.1f} MB"


def main():
    args = sys.argv[1:]
    if not args:
        fail("usage: attach-commit <sha> [CO-N]")
    sha_arg = args[0]
    key_arg = args[1] if len(args) > 1 else None

    # Validate the sha and normalise it to the full hash.
    sha = git(["rev-parse", "--verify", f"{sha_arg}^{{commit}}"]).strip()

    message = git(["show", "-s", "--format=%B", sha]).rstrip("\n")
    committed_at = git(["show", "-s", "--format=%cI", sha]).strip()
    author_name = git(["show", "-s", "--format=%an", sha]).strip()

    key = key_arg or parse_key(message)
    if not key:
        fail(
            f"no card key found in the message of {sha[:8]}; "
            f"pass one explicitly: attach-commit {sha_arg} CO-N"
        )

    stat = git(["show", sha, "--stat", "--format="]).strip("\n")
    raw_diff = git(["show", sha, "--format=", "--patch"])
    diff = prune_diff(raw_diff)

    file_count = len(re.findall(r"^diff --git ", raw_diff, re.MULTILINE))

    payload = json.dumps(
        {
            "sha": sha,
            "message": message,
            "stat": stat,
            "diff": diff,
            "committedAt": committed_at,
            "authorName": author_name,
        }
    ).encode("utf-8")

    req = urllib.request.Request(
        f"{API_URL}/cards/{key}/commits",
        data=payload,
        headers=with_token({"Content-Type": "application/json"}),
        method="POST",
    )
    try:
        with urllib.request.urlopen(req) as res:
            res.read()
    except urllib.error.HTTPError as err:
        body = err.read().decode("utf-8", "replace")
        fail(f"API responded {err.code}: {body}")
    except urllib.error.URLError as err:
        fail(f"could not reach the API at {API_URL} ({err.reason}). Is it running?")

    diff_bytes = len(diff.encode("utf-8"))
    print(f"✓ {key} {sha[:8]} — {file_count} file(s), {human_bytes(diff_bytes)} diff → attached")


if __name__ == "__main__":
    main()

File: plugins/claude-organizer/skills/implement/scripts/attach-commit.mjs
#!/usr/bin/env node
// Capture a commit's diff and attach it to its card via the REST API.
//
//   node scripts/attach-commit.mjs <sha> [CO-N]
//   pnpm attach-commit <sha> [CO-N]
//
// The card key is parsed from the commit message (`feat(...): … (CO-12)`) unless
// passed explicitly. The diff goes straight from `git` to the API over HTTP — it
// never passes through an AI context (no MCP, no tokens spent reading it).
//
// Zero dependencies: standalone Node 18+ (global fetch). A Python twin lives at
// scripts/attach-commit.py for machines without Node. Keep the two in sync.
//
// Config: CO_API_URL (default http://127.0.0.1:4400).

import { execFileSync } from 'node:child_process'

// Windows' cp1252 console can't encode the ✓/✗/→ glyphs; force UTF-8 so the
// final log doesn't crash after the POST already succeeded.
process.stdout.setDefaultEncoding?.('utf8')
process.stderr.setDefaultEncoding?.('utf8')

const API_URL = (process.env.CO_API_URL || 'http://127.0.0.1:4400').replace(/\/$/, '')

// Card-scoped token minted by the MCP (issue_commit_token); only needed when the
// API has auth on. Absent in sem-auth mode — then no extra header is sent.
const COMMIT_TOKEN = process.env.CO_COMMIT_TOKEN

function withToken(headers = {}) {
  return COMMIT_TOKEN
    ? { ...headers, 'X-CO-Commit-Token': COMMIT_TOKEN }
    : headers
}

// Files whose body is noise: store the header + a note instead of the patch.
const LOCKFILES = new Set([
  'pnpm-lock.yaml',
  'package-lock.json',
  'yarn.lock',
  'npm-shrinkwrap.json',
  'bun.lockb'
])
// Per-file line cap; beyond it the patch is truncated with a note.
const MAX_LINES_PER_FILE = 1000

// Raster images are captured as before/after attachments and rendered in the web
// diff (CO-392); other binaries keep the plain note. Mirrors the attachment
// allow-list in @claude-organizer/shared.
const IMAGE_EXT_MIME = {
  png: 'image/png',
  jpg: 'image/jpeg',
  jpeg: 'image/jpeg',
  gif: 'image/gif',
  webp: 'image/webp'
}
// Skip the upload above this and keep the binary note; aligned with the API's
// 10 MB upload ceiling so we never ship a blob the server would reject.
const MAX_IMAGE_BYTES = 10 * 1024 * 1024

function fail(msg) {
  console.error(`✗ ${msg}`)
  process.exit(1)
}

function imageMime(path) {
  return IMAGE_EXT_MIME[(path.split('.').pop() || '').toLowerCase()] || null
}

// The sentinel that replaces `Binary files … differ` so the web can render the
// image. Must match buildDiffImageSentinel / DIFF_IMAGE_SENTINEL_PREFIX in
// @claude-organizer/shared (a side is omitted when absent).
function imageSentinel(oldId, newId) {
  let line = '# image'
  if (oldId) line += ` old=${oldId}`
  if (newId) line += ` new=${newId}`
  return line
}

// Binary-safe git read (no utf8 decode); null when the path is absent at that ref
// (an added file has no parent blob; a deleted file has none at the commit).
// stderr is ignored — a missing blob is an expected miss, not an error to print.
function gitBlob(args) {
  try {
    return execFileSync('git', args, {
      maxBuffer: 256 * 1024 * 1024,
      stdio: ['ignore', 'pipe', 'ignore']
    })
  } catch {
    return null
  }
}

// Upload one image side to the CO-393 endpoint; returns its att id, or null when
// the blob is missing/over the cap, OR the upload errored — a per-image failure
// degrades to the plain binary note for that section, never discarding the whole
// diff (the diff POST below stays the hard gate on a down API).
async function uploadImage(buf, path, key) {
  if (!buf || !buf.length || buf.length > MAX_IMAGE_BYTES) return null
  const mime = imageMime(path)
  if (!mime) return null
  const form = new FormData()
  form.append('file', new Blob([buf], { type: mime }), path.split('/').pop() || 'image')
  try {
    const res = await fetch(
      `${API_URL}/cards/${encodeURIComponent(key)}/commit-images`,
      { method: 'POST', headers: withToken(), body: form }
    )
    return res.ok ? (await res.json()).id : null
  } catch {
    return null
  }
}

function git(args) {
  try {
    return execFileSync('git', args, {
      encoding: 'utf8',
      maxBuffer: 256 * 1024 * 1024
    })
  } catch (err) {
    fail(`git ${args.join(' ')} failed: ${err.message.split('\n')[0]}`)
  }
}

// `CO-12`, `ABC-7` — an uppercase prefix, a dash, digits. We prefer the last
// match on the subject line (the project convention puts the card's own key at
// the end of the subject), then fall back to the whole message.
const KEY_RE = /\b([A-Z][A-Z0-9]*-\d+)\b/g

function parseKey(message) {
  const subject = message.split('\n', 1)[0]
  const onSubject = [...subject.matchAll(KEY_RE)].map(m => m[1])
  if (onSubject.length) return onSubject[onSubject.length - 1]
  const anywhere = [...message.matchAll(KEY_RE)].map(m => m[1])
  return anywhere.length ? anywhere[0] : null
}

function isIgnored(path) {
  const base = path.split('/').pop() || ''
  if (LOCKFILES.has(base)) return true
  if (/\.min\.[^.]+$/.test(base)) return true
  if (path.startsWith('dist/') || path.includes('/dist/')) return true
  return false
}

function sectionPath(lines) {
  for (const l of lines) if (l.startsWith('+++ b/')) return l.slice(6)
  for (const l of lines) if (l.startsWith('--- a/')) return l.slice(6)
  const m = lines[0].match(/^diff --git a\/(.+) b\/(.+)$/)
  return m ? m[2] : ''
}

// Capture an image binary section: upload its old/new blobs and replace the
// binary marker (and any GIT binary patch body) with the image sentinel. Falls
// back to the original lines (plain note) when neither side could be uploaded.
async function captureImageSection(lines, path, key, sha) {
  const oldId = await uploadImage(gitBlob(['show', `${sha}^:${path}`]), path, key)
  const newId = await uploadImage(gitBlob(['show', `${sha}:${path}`]), path, key)
  if (!oldId && !newId) return lines.join('\n')
  const binIdx = lines.findIndex(
    l => /^Binary files .* differ$/.test(l) || l.startsWith('GIT binary patch')
  )
  const head = binIdx >= 0 ? lines.slice(0, binIdx) : lines
  return [...head, imageSentinel(oldId, newId)].join('\n')
}

// Split the unified patch per file and apply the safeguards.
async function pruneDiff(patch, key, sha) {
  if (!patch.trim()) return ''
  const sections = []
  let current = null
  for (const line of patch.split('\n')) {
    if (line.startsWith('diff --git ')) {
      if (current) sections.push(current)
      current = [line]
    } else if (current) {
      current.push(line)
    }
  }
  if (current) sections.push(current)

  return (await Promise.all(sections
    .map(async (lines) => {
      const path = sectionPath(lines)
      const isBinary = lines.some(
        l => /^Binary files .* differ$/.test(l) || l.startsWith('GIT binary patch')
      )
      if (isBinary) {
        if (!isIgnored(path) && imageMime(path)) {
          return captureImageSection(lines, path, key, sha)
        }
        return lines.join('\n')
      }
      if (isIgnored(path)) {
        return `${lines[0]}\n# diff omitted (lockfile/generated)`
      }
      if (lines.length > MAX_LINES_PER_FILE) {
        const omitted = lines.length - MAX_LINES_PER_FILE
        return (
          lines.slice(0, MAX_LINES_PER_FILE).join('\n')
          + `\n# ... (truncated: ${omitted} lines omitted)`
        )
      }
      return lines.join('\n')
    }))).join('\n')
}

function humanBytes(n) {
  if (n < 1024) return `${n} B`
  if (n < 1024 * 1024) return `${(n / 1024).toFixed(1)} KB`
  return `${(n / 1024 / 1024).toFixed(1)} MB`
}

async function main() {
  const [shaArg, keyArg] = process.argv.slice(2)
  if (!shaArg) {
    fail('usage: attach-commit <sha> [CO-N]')
  }

  // Validate the sha and normalise it to the full hash.
  const sha = git(['rev-parse', '--verify', `${shaArg}^{commit}`]).trim()

  const message = git(['show', '-s', '--format=%B', sha]).replace(/\n+$/, '')
  const committedAt = git(['show', '-s', '--format=%cI', sha]).trim()
  const authorName = git(['show', '-s', '--format=%an', sha]).trim()

  const key = keyArg || parseKey(message)
  if (!key) {
    fail(
      `no card key found in the message of ${sha.slice(0, 8)}; pass one explicitly: attach-commit ${shaArg} CO-N`
    )
  }

  const stat = git(['show', sha, '--stat', '--format=']).replace(/^\n+/, '').replace(/\n+$/, '')
  const rawDiff = git(['show', sha, '--format=', '--patch'])
  const diff = await pruneDiff(rawDiff, key, sha)

  const fileCount = (rawDiff.match(/^diff --git /gm) || []).length

  const res = await fetch(`${API_URL}/cards/${encodeURIComponent(key)}/commits`, {
    method: 'POST',
    headers: withToken({ 'Content-Type': 'application/json' }),
    body: JSON.stringify({ sha, message, stat, diff, committedAt, authorName })
  }).catch((err) => {
    fail(`could not reach the API at ${API_URL} (${err.message}). Is it running?`)
  })

  if (!res.ok) {
    const body = await res.text()
    fail(`API responded ${res.status}: ${body}`)
  }

  console.log(
    `✓ ${key} ${sha.slice(0, 8)} — ${fileCount} file(s), ${humanBytes(Buffer.byteLength(diff, 'utf8'))} diff → attached`
  )
}

main()
